geoutils/georaster.py

```diff
@@ -526,6 +527,96 @@ def shift(self, xoff, yoff):
+            print('Several raster band: using nodata value for all bands')
+            ndv = [ndv]*self.count
+        elif isinstance(ndv,list) and self.count == 1:
+            print('Only one raster band: using first nodata value provided')
+            ndv = ndv[0]
+
+        meta = self.ds.meta
+        imgdata = self.data
+        pre_ndv = self.nodata
+
+        meta.update({'nodata': ndv})
+
+        if update_array and pre_ndv is not None:
+            #nodata values are specific to each band
+
+            #tried to do data cube at once, doesn't work :(
+            # ind = (imgdata == np.array([pre_ndv])[:, None, None] * np.ones((self.ds.height, self.ds.width))[None, :, :])
+            # output = np.array([ndv])[:, None, None] * np.ones((self.ds.height, self.ds.width))[None, :, :]
+            # imgdata[ind] = output
+
+            #let's do a loop then
+            if self.count == 1:
+                ind = imgdata[:] == pre_ndv
+                imgdata[ind] = ndv
+            else:
+                for i in range(self.count):
+                    ind = imgdata[i,:] == pre_ndv[i]
+                    imgdata[i,ind] = ndv[i]
+        else:
+            imgdata = None
+
+        self._update(metadata=meta,imgdata=imgdata)
+
+    def set_dtypes(self,dtypes,update_array=True):
+
+        """
+        Set new dtypes for bands (and possibly update arrays)
+
+        :param dtypes: data types
+        :type dtypes:  list or type or str
+        :param update_array: change the existing dtype in arrays
+        :type: update_array: bool
+        """
+
+        if not (isinstance(dtypes,list) or isinstance(dtypes, type) or isinstance(dtypes, str)):
+            raise ValueError(
+                "Type of dtypes not understood, must be list or type or str")
+        elif isinstance(dtypes, type) or isinstance(dtypes, str):
+            print('Several raster band: using data type for all bands')
+            dtypes = (dtypes,) * self.count
+        elif isinstance(dtypes, list) and self.count == 1:
+            print('Only one raster band: using first data type provided')
+            dtypes = tuple(dtypes)
+
+        meta = self.ds.meta
+        imgdata = self.data
+
+        #for rio.DatasetReader.meta, the proper name is "dtype"
+        meta.update({'dtype': dtypes[0]})
+
+        #this should always be "True", as rasterio doesn't change the array type by default:
+        #ValueError: the array's dtype 'int8' does not match the file's dtype 'uint8'
+        if update_array:
+            if self.count == 1:
+                imgdata = imgdata.astype(dtypes[0])
+            else:
+                #TODO: double-check, but I don't think we can have different dtypes for bands with rio (1 dtype in meta)
+                imgdata = imgdata.astype(dtypes[0])
+                for i in imgdata.shape[0]:
+                    imgdata[i,:] = imgdata[i,:].astype(dtypes[0])
+        else:
+            imgdata = None
```

Comment:
If the data was loaded, your method will remove it. Should imgdata just not be changed here?

Reply:
I'm not sure it will, as I just pass a None argument to the __update() method which relinks to the data loaded:
>         if imgdata is None:
>             imgdata = self.data
And updating imgdata doesn't update self.data?

Reply:
Alright, I didn't realize that _update's default was None and that by default it would set it to self.data. In that case, it's all good!